src/main/java/org/truffleruby/core/array/ArrayDupNode.java

```diff
@@ -38,25 +38,19 @@
     public DynamicObject dupProfiledSize(DynamicObject from,
             @Cached("of(from)") ArrayStrategy strategy,
             @Cached("strategy.getSize(from)") int cachedSize) {
-        return copyExplode(from, strategy, cachedSize);
+        return copyArray(from, strategy, cachedSize);
     }
 
-    @ExplodeLoop
-    private DynamicObject copyExplode(DynamicObject from, ArrayStrategy strategy, int cachedSize) {
-        final ArrayMirror mirror = strategy.newMirror(from);
-        final ArrayMirror copy = strategy.newArray(cachedSize);
-        for (int i = 0; i < cachedSize; i++) {
-            copy.set(i, mirror.get(i));
-        }
-        return allocateArray(coreLibrary().getArrayClass(), copy.getArray(), cachedSize);
+    private DynamicObject copyArray(DynamicObject from, ArrayStrategy strategy, int Size) {
+        final ArrayMirror copy = strategy.makeStorageShared(from).extractRange(0, Size);
+        return allocateArray(coreLibrary().getArrayClass(), copy.getArray(), Size);
     }
```

Comment:
I want to keep the `@ExplodeLoop` here, it's important for performance notably in Optcarrot.

Reply:
It's no longer a loop to explode, at least not here. The original array is converted to shared storage and then `extractRange` simply creates a second pointer into that storage. If we want to explode loops it should probably be when we convert shared storaged to unshared storage, and possibly just use s simple copy if the array is small rather then making shared storage.

I'll run optcarrot with a little instrumentation to see how often it ends up unsharing storage. Initial performance doesn't seem to bad, as we achieve stable 60fps, but I'll take a closer look.

Reply:
IIRC, the case that matters for this is when splatting arguments and RubyArguments.pack() get called.
There it's important that the Object[]'s `length` is a compile-time constant, otherwise we end up with a slow arraycopy.
Arguably, profiling the number of arguments should be done in `RubyArguments.pack()`, transformed to a node, to avoid arraycopy with unknown length.

The main occurrence of this pattern in Optcarrot is https://github.com/oracle/truffleruby/blob/355c1abdb06087be04e0ce7f250c3d32e7f61806/bench/optcarrot/lib/optcarrot/cpu.rb#L939